src/jdk.jartool/share/classes/sun/tools/jar/Validator.java

```diff
@@ -62,20 +82,139 @@ final class Validator {
     private Set<String> concealedPkgs = Collections.emptySet();
     private ModuleDescriptor md;
     private String mdName;
+    private final ZipInputStream zis;
 
-    private Validator(Main main, ZipFile zf) {
+    private Validator(Main main, ZipFile zf, ZipInputStream zis) {
         this.main = main;
         this.zf = zf;
+        this.zis = zis;
         checkModuleDescriptor(MODULE_INFO);
     }
 
-    static boolean validate(Main main, ZipFile zf) throws IOException {
-        return new Validator(main, zf).validate();
+    static boolean validate(Main main, ZipFile zf, ZipInputStream zis) throws IOException {
+        return new Validator(main, zf, zis).validate();
+    }
+
+    /**
+     * Validate that the CEN/LOC file name header field adheres  to
+     * PKWARE APPNOTE-6.3.3.TXT:
+     *
+     * 4.4.17.1 The name of the file, with optional relative path.
+     * The path stored MUST not contain a drive or
+     * device letter, or a leading slash.  All slashes
+     * MUST be forward slashes '/' as opposed to
+     * backwards slashes '\' for compatibility with Amiga
+     * and UNIX file systems etc.
+     * Also validate that the file name is not "." and that any name element is
+     * not equal to ".."
+     *
+     * @param entryName CEN/LOC header file name field entry
+     * @return true if a valid Zip Entry file name; false otherwise
+     */
+    public static boolean isZipEntryNameValid(String entryName) {
+        return !INVALID_ZIP_ENTRY_NAME_PATTERN.matcher(entryName).find();
+    }
+
+    /**
+     * Helper class to deduplicate entry names.
+     * Keep a coutner for duplicate entry names and check if the entry name is valid on first add.
+     */
+    private class DedupEntryNames {
+        LinkedHashMap<String, Integer> entries = new LinkedHashMap<>();
+        boolean isCEN;
+
+        public DedupEntryNames(boolean isCEN) {
+            this.isCEN = isCEN;
```

Comment:
Might be clearer if this was an enum vs boolean?

Reply:
I feel the same way but wondering if that's an overkill.

Reply:
Refactored to consolidate validation on entries into a standalone EntryValidator. Hope that's easier to understand.